src/support/generation.c

```diff
@@ -156,25 +156,30 @@ __wt_gen_drain(WT_SESSION_IMPL *session, int which, uint64_t generation)
                 __wt_sleep(0, 10);
 
             /*
-             * If we wait for more than a minute, log the event. In DIAGNOSTIC mode, abort if we
-             * ever wait more than 3 minutes, that's forever.
+             * If we wait for more than a minute, log the event. In diagnostic mode, abort if we
+             * ever wait more than the configured timeout.
              */
             if (minutes == 0) {
                 minutes = 1;
                 __wt_epoch(session, &start);
             } else {
                 __wt_epoch(session, &stop);
                 time_diff_ms = WT_TIMEDIFF_MS(stop, start);
-#define WT_GEN_DRAIN_TIMEOUT_MIN 4
+
                 if (time_diff_ms > minutes * WT_MINUTE * WT_THOUSAND) {
                     __wt_verbose_notice(session, WT_VERB_GENERATION,
                       "%s generation drain waited %" PRIu64 " minutes", __gen_name(which), minutes);
                     ++minutes;
-                    WT_ASSERT(session, minutes < WT_GEN_DRAIN_TIMEOUT_MIN);
                 }
-                /* Enable extra logs 20ms before timing out. */
-                else if (!verbose_timeout_flags &&
-                  time_diff_ms > (WT_GEN_DRAIN_TIMEOUT_MIN * WT_MINUTE * WT_THOUSAND - 20)) {
+
+                /* If no timeout is configured, there is nothing else to do. */
+                if (conn->gen_drain_timeout_ms == 0)
+                    continue;
+
+                /* Enable extra logs 20ms before reaching the timeout. */
+                if (!verbose_timeout_flags &&
```

Comment:
I can see a potential issue here if diagnostic mode is **disabIed**: we may enable those flags and logs will be filled with new messages even though we won't abort because diagnostic is disabled.

We could potentially:

- Disable those logs when the resource has drained
- Only enable those logs in diagnostic mode

Reply:
It's a reasonable point. Could you add a Python test that configures this vluae very small, and maybe we can use that to determine what the right choice is?

I'm a bit inclined to output these messages for 20ms leading up to the timeout and then disable them again if the assertion doesn't fire. It might be fraught to reset the verbose levels back - would we need to save original values to do that? If it's that complex to turn them back off, I'd probably be inclined to not increase the verbosity level in diagnostic builds.

Reply:
I went for the second approach as this code may be revisited in [WT-11097](https://jira.mongodb.org/browse/WT-11097). See 6dc5990b9eeae665ba96692229c078731fa6fd63.